source/common/http/codec_client.cc

```diff
@@ -138,6 +138,19 @@ void CodecClient::onData(Buffer::Instance& data) {
       host_->cluster().stats().upstream_cx_protocol_error_.inc();
     }
   }
+
+  // The HTTP/1 codec will pause dispatch after a single response is complete. If there is more
+  // data, drain and close the connection.
+  if (codec_->protocol() < Protocol::Http2 && data.length() > 0) {
+    ENVOY_CONN_LOG(debug, "CONNECTION STATE {}", *connection_, connection_->state());
+    ENVOY_CONN_LOG(debug, "Draining remaining {} bytes", *connection_, data.length());
+    host_->cluster().stats().upstream_cx_protocol_error_.inc();
+    data.drain(data.length());
+    close();
+  }
+
+  // All data should be consumed at this point.
+  // ASSERT(data.length() == 0);
```

Comment:
nit: remove commented out code you were using for debugging.  The most likely explanation for data remaining after dispatch is an error that already triggered a connection close.

Also, I wonder if tests succeed if you change the ASSERT to:
ASSERT(data.length() == 0 || connection_->state() != State::Open);

Reply:
Oof, thanks. I was experimenting with connection state, realized some of the breakages from this were test issues.